activerecord/lib/active_record/core.rb

```diff
@@ -79,6 +79,8 @@ def self.configurations
 
       class_attribute :strict_loading_by_default, instance_accessor: false, default: false
 
+      class_attribute :strict_association_name, instance_writer: false, default: false
```

Comment:
I think we could go with a single `ActiveRecord.allow_deprecated_singular_assocaition_name` (or whatever name you think is good, I didn't think much about it).

Reply:
Surprised the rails spellchecker complains about `association`, but I like `allow_deprecated_singular_assocaitions_name` enough I went with it. Thanks!

Reply:
Sorry it wasn't clear. Aside from the name, I was suggesting to make it an `ActiveRecord` singleton attribute:

```ruby
module ActiveRecord
  singleton_class.attr_accessor :allow_deprecated_singular_assocaitions_name
end
```

This means it's no longer inheritable (but the value here is limited), but you can directly check it where it's needed and don't have to pass it around as much. That should significantly cut down the diff.